watch-library/hardware/watch/watch_private.c

```diff
@@ -245,43 +245,40 @@ void _watch_enable_usb(void) {
 
 // this function ends up getting called by printf to log stuff to the USB console.
 int _write(int file, char *ptr, int len) {
+    int written = 0;
     (void)file;
-    if (hri_usbdevice_get_CTRLA_ENABLE_bit(USB)) {
-        tud_cdc_n_write(0, (void const*)ptr, len);
-        tud_cdc_n_write_flush(0);
-        return len;
+
+    // It's important that this call blocks, as returning 0 bytes
+    // written causes newlib to fail the flush and throw
+    // away any data in the buffer.
+    // Note that it's fine not to write all the bytes, it's just
+    // that we have to write _something_.
+    while (tud_cdc_connected() && written == 0) {
```

Comment:
I'm now wondering if it's a reasonable approach to just try to do this for some amount of time _then_ fail to avoid deadlocks. Of course, I could put an intermediate buffer in, but tiny USB already has a buffer here (which we could increase...).